### PR TITLE
New DataPartitionType DATA

### PR DESCRIPTION
**Pull request by apoorvtintin**
Increases memory efficiency during large scale training, input batches and labels are sharded along the 'data' axis. 
Added new input data sharding option ```DataPartitionType.DATA```.

### CODE REVIEW

File: axlearn/common/utils.py [hunk 1]
```diff
@@ -499,19 +499,29 @@ def add_leaves(i, x):
...
 
-def input_partition_spec() -> PartitionSpec:
-    """Returns partition spec for the input batch.
+class DataPartitionType(Enum):
+    # Data are fully partitioned across all devices.
+    FULL = "full"
+    # Data are fully replicated across all devices.
+    REPLICATED = "replicated"
+    # Data are partially partitioned across data axis
+    DATA = "data"
```
[the remark this answers is not in the document]
Comment (ptoulme-aws) [reply axlearn/common/utils.py:508]: DATA replicates over the sequence dimension. so the spec is ("data", None) versus ("data", "model") for FULL

**Review by ruomingp — COMMENTED** [review]
> Increases memory efficiency

Do you have measurements on how DATA improves memory efficiency? Thanks.

**Comment by ptoulme-aws** [comment]
> > Increases memory efficiency
> 
> Do you have measurements on how DATA improves memory efficiency? Thanks.



> > Increases memory efficiency
> 
> Do you have measurements on how DATA improves memory efficiency? Thanks.

By replicating the sequence length over TP workers we limit collectives and dynamic-slices introduced by the SPMD partitioner. This lowers overall step time and also allows us to run sequence parallelism over TP workers.

**Comment by ruomingp** [comment]
> > > Increases memory efficiency
> > 
> > 
> > Do you have measurements on how DATA improves memory efficiency? Thanks.
> 
> > > Increases memory efficiency
> > 
> > 
> > Do you have measurements on how DATA improves memory efficiency? Thanks.
> 
> By replicating the sequence length over TP workers we limit collectives and dynamic-slices introduced by the SPMD partitioner. This lowers overall step time and also allows us to run sequence parallelism over TP workers.

Thanks. Do you have quantitative measurments?

**Comment by ptoulme-aws** [comment]
> > > > Increases memory efficiency
> > > 
> > > 
> > > Do you have measurements on how DATA improves memory efficiency? Thanks.
> > 
> > 
> > > > Increases memory efficiency
> > > 
> > > 
> > > Do you have measurements on how DATA improves memory efficiency? Thanks.
> > 
> > 
> > By replicating the sequence length over TP workers we limit collectives and dynamic-slices introduced by the SPMD partitioner. This lowers overall step time and also allows us to run sequence parallelism over TP workers.
> 
> Thanks. Do you have quantitative measurements?

No, we do not. It is more when we inspect the HLO after SPMD partition pass we see much more optimal sharding. Less all-to-alls and less dynamic-slices on right hand side.

**Pull request closed without merging**